src/upload.rs

```diff
@@ -0,0 +1,215 @@
+        .args([
+            "-av",
+            "--prune-empty-dirs",
+            "--include=*/",
+            "--include=Forc.toml",
+            "--include=Forc.lock",
+            "--include=*.sw",
+            "--exclude=*",
+            "unpacked/",
+            "project",
+        ])
+        .current_dir(upload_dir)
+        .output()
+        .expect("Failed to copy project files");
+
+    if !output.status.success() {
+        return Err(UploadError::CopyFiles);
+    }
+
+    // Pack the new tarball.
+    let final_tarball_path = upload_dir.join("project.tgz");
+    let tar_gz = File::create(&final_tarball_path).unwrap();
+    let enc = GzEncoder::new(tar_gz, Compression::default());
+    let mut tar = tar::Builder::new(enc);
+
+    // Add files to the tar archive
+    tar.append_dir_all(".", &project_dir).unwrap();
+
+    // Finish writing the tar archive
+    tar.finish().unwrap();
+
+    // Make sure the GzEncoder finishes and flushes all data
+    let enc = tar.into_inner().unwrap();
+    enc.finish().unwrap();
+
+    // Store the tarball in IPFS.
+    let tarball_ipfs_hash = pinata_client
+        .upload_file_to_ipfs(&final_tarball_path)
+        .await?;
+
+    fn find_file_in_dir_by_suffix(dir: &Path, suffix: &str) -> Option<PathBuf> {
+        let dir = fs::read_dir(dir).ok()?;
+        dir.flatten()
+            .filter_map(|entry| {
+                let path = entry.path();
+                if path.is_file() {
+                    if let Some(file_name) = path.file_name() {
+                        if let Some(file_name_str) = file_name.to_str() {
+                            // Check if the path is a file and ends with the suffix
+                            if file_name_str.ends_with(suffix) {
+                                return Some(path); // Return the first found file
+                            }
+                        }
+                    }
+                }
+                None
+            })
+            .next()
+    }
+
+    // Store the ABI in IPFS.
+    let abi_ipfs_hash = match find_file_in_dir_by_suffix(&release_dir, "-abi.json") {
+        Some(abi_path) => Some(pinata_client.upload_file_to_ipfs(&abi_path).await?),
+        None => None,
+    };
+
+    // Generate the bytecode identifier and store in the database along with the ABI hash.
+    let bytecode_identifier = match find_file_in_dir_by_suffix(&release_dir, ".bin") {
+        Some(bin_path) => Some(
+            get_bytecode_id(&bin_path).map_err(|err| UploadError::BytecodeId(err.to_string()))?,
+        ),
+        None => None,
+    };
+
+    let upload = NewUpload {
+        id: *upload_id,
+        source_code_ipfs_hash: tarball_ipfs_hash,
+        forc_version,
+        abi_ipfs_hash,
+        bytecode_identifier,
+    };
+
+    Ok(upload)
+}
+
+/// Installs the given version of forc at the specific root path using cargo-binstall.
+pub fn install_forc_at_path(forc_version: &str, forc_path: &Path) -> Result<(), UploadError> {
```

Comment:
Instead of using the CLI; i'd reccomend using library/SDK functions to perform uploading operations.
They would be much lighter and ephemeral as one would not need to have the binary installed.

Reply:
Are you saying to use forc-pkg to build the sway project rather than using the forc CLI? Or are you suggesting using a library to install forc? If you know of libraries that can do this I'd be interested, I looked around but couldn't find one.

Reply:
Yea i was saying we shouldn't rely on the user having forc-binaries installed; implying requiring external deps.
Note: Could be a future PR though as an improvement.

Reply:
I might be a little lost here but isn't this a server that we are running? What users are we referring to exactly?

My understanding was that: we run this service, once a user want's to actually publish a project, `forc upload` (or whatever the cli command will be) will make an API call into this service uploading the project. 

Then the server will get the project, and extract the `bytecode identifier` from it (bytecode identifier in this context would basically be an id generated from the existing bincode that "generalizes" the project, meaning that the effect of different initial configurable values etc are removed to get an id that describes only the "core" of the contract)

Reply:
@kayagokalp has it right. This isn't expecting a user to have `forc` binary installed, it's installing it on the server (running in docker) and using it to build the sway project that it received via the upload API.

We wouldn't be able to use forc-pkg since we have to support any version of forc.